src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Serialization/RazorProjectInfoDeserializer.cs

```diff
@@ -14,6 +15,21 @@ private RazorProjectInfoDeserializer()
     {
     }
 
+    public RazorProjectInfo? DeserializeFromString(string? projectInfoBase64String)
+    {
+        RazorProjectInfo? razorProjectInfo = null;
+
+        // ProjectInfo will be null if project is being deleted and should be removed
+        if (projectInfoBase64String is not null)
+        {
+            var projectInfoBytes = Convert.FromBase64String(projectInfoBase64String);
+            using var stream = new MemoryStream(projectInfoBytes);
+            razorProjectInfo = DeserializeFromStream(stream);
+        }
+
+        return razorProjectInfo;
+    }
```

Comment:
💭 Performance: Instead of producing the bytes and then copying all of the bytes to a stream before deserializing, you could avoid the copy by adding a new `RazorProjectInfo.DeserializeFrom(ReadOnlyMemory<byte>)` overload that calls the `MessagePackSerializer.Deserialize(ReadOnlyMemory<byte>)` overload:

![image](https://github.com/dotnet/razor/assets/116161/c87c72ba-7e39-47a0-bcd5-87269d7b3347)

Then you should be to pass to `projectInfoBytes` directly.

Reply:
Note: There are similar changes that could be made to the serialization side in `IProjectSnapshoitExtensions` below to use an `IBufferWriter<byte>` to avoid copying arrays around into and out of memory streams. Maybe file an issue for this to fix up later? I'm happy to take a look at that.

Reply:
We will likely do serialization differently (see David's comment in the original PR) by directly serializing RazorProjectInfo instead of writing it to a binary and converting that to a string. I'll skip doing those improvements now and we can chat about it separately in the "improved serialization" PR.

Reply:
FWIW, serializing RazorProjectInfo directly may not be _quite_ the right approach perf-wise -- at least for the `ProjectWorkspaceState`. `ProjectWorkspaceState` contains an array of `TagHelperDescriptors`, which can grow huge. We probably want to be a bit smarter about passing checksums for any tag helpers. I imagine it would work like this:

1. Client sends request `RazorProjectInfo` with checksums baked in for `TagHelperDescriptors`. It awaits a response from the server.
2. Server receives the request and uses the checksums to determine which `TagHelperDescriptors` it already knows about and which ones it needs. If it needs any tag helpers, it passes back the checksums for the missing tag helpers in its response.
3. The client receives the response. If there are any checksums for tag helpers needed by the server, it passes those to the server via another endpoint.

The plan @davidwengier suggested is good for now, but it's going to result in a lot of extra serialization cost. Long term, I don't think we'll want the perf hit of fully passing `RazsorProjectInfo`. That'd likely go away with co-hosting but we're not there yet.

_Just thinking out loud -- no changes requested here!_ :smile: